frontend/src/pages/projects/screens/spawner/useNotebookSizeState.ts

```diff
@@ -0,0 +1,32 @@
+import React from 'react';
+import useNotebookDeploymentSize from '~/pages/projects/screens/detail/notebooks/useNotebookDeploymentSize';
+import { NotebookKind } from '~/k8sTypes';
+import { NotebookSize } from '~/types';
+import { getCustomNotebookSize } from '~/pages/projects/utils';
+import { useNotebookSize } from './useNotebookSize';
+
+export const useNotebookSizeState = (
+  existingNotebook: NotebookKind | undefined,
+): {
+  selectedSize: NotebookSize;
+  setSelectedSize: (size: string) => void;
+  sizes: NotebookSize[];
+} => {
+  const { size: notebookSize } = useNotebookDeploymentSize(existingNotebook);
+  const { selectedSize, setSelectedSize, sizes } = useNotebookSize();
+
+  const CUSTOM_NOTEBOOK_SIZE = getCustomNotebookSize(existingNotebook);
+
+  React.useEffect(() => {
+    if (notebookSize) {
+      setSelectedSize(notebookSize.name);
+    } else {
+      if (existingNotebook && !sizes.some((size) => size.name === CUSTOM_NOTEBOOK_SIZE.name)) {
+        sizes.unshift(CUSTOM_NOTEBOOK_SIZE);
+      }
+      setSelectedSize(CUSTOM_NOTEBOOK_SIZE.name);
+    }
+  }, [notebookSize, setSelectedSize, sizes, existingNotebook, CUSTOM_NOTEBOOK_SIZE]);
```

Comment:
Every time i see a `useEffect` without any async behavior it's likely possibly to do the same using `useMemo` and properly assigning the default state when `useState` is first defined. In this particular case the default selected size comes from `useNotebookSize` and assigns `sizes[0]` which isn't correct. It will not always be the first in the list and a good solution would be to accept the default state as a parameter to the hook.

Reply:
This fixes the [testing issue](https://github.com/opendatahub-io/odh-dashboard/pull/2944#discussion_r1671997549) too.